### PR TITLE
EDS: Handle missing ebook URL

### PR DESCRIPTION
**Pull request by maccabeelevine**
The EDS record driver method `getEbookLink` returns false if there is no appropriate link found.  The EDSrecordController method `redirectToEbook` does not account for this possibility and just attempts a redirect to the returned URL.  This ultimately results in a laminas InvalidUriException when it tries to parse the boolean value as a URL.  That throws up an ugly full-screen error message.

This is a small-scope fix which just throws a VuFind-specific error message when the link is not found.  This is displayed in the usual VuFind UI with a human-readable error.

### CODE REVIEW

File: module/VuFind/src/VuFind/Controller/EdsrecordController.php [hunk 1]
```diff
@@ -80,7 +80,11 @@ protected function redirectToEbook($format, $method)
             }
             throw new ForbiddenException('Access denied.');
         }
-        return $this->redirect()->toUrl($driver->tryMethod($method));
+        $url = $driver->tryMethod($method);
+        if (!$url) {
+            throw new \VuFind\Exception\BadConfig('EDS API did not provide a URL for this record.');
```
Comment (maccabeelevine) [inline module/VuFind/src/VuFind/Controller/EdsrecordController.php]: There is probably a better exception class to use for this.  I would argue it *is* caused by bad configuration, specifically on the EDS side, which is effectively what I'm saying in the exception message.  But the class may have been intended for only local config errors.

Comment (maccabeelevine) [reply module/VuFind/src/VuFind/Controller/EdsrecordController.php]: You could also argue that it should not throw an exception at all, but should redirect to something more useful.  But I'm not sure what that would be and likely depends on the (unknown) context of why the URL lookup failed in the first place.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Would it be better still if you displayed a more informative error message in the user interface by setting a flash message and redirecting to the impacted record page? For example, instead of throwing the exception, do this:

```
        if (!$url) {
            $this->flashMessenger()->addErrorMessage('EDS API did not provide a URL for this record.');
            return $this->redirect()->toRoute('edsrecord', ['id' => $this->params()->fromRoute('id')]);
        }
```

If we go this route, though, we should probably replace `EDS API did not provide a URL for this record.` with something more generic, like `error_accessing_full_text`, and then we could put something in en.ini like:

```
error_accessing_full_text = "There was an error retrieving the full text you requested."
```

What do you think?

**Comment by maccabeelevine** [comment]
@demiankatz I took your idea, that's certainly much friendlier.  There is a tradeoff here in that it ends up looking stupid, because there actually *is* a link that works to get the full text, right under the thumbnail, about 100px below the error message.

image

But that is from the record driver's `getPLink()` not `getLinkedFullTextLink()`.  So another obvious solution would be to have the latter fallback on the former, but I have no idea what a "p link" is and I have no confidence it would not be ok -- even though in this case it appears to work.

**Review by demiankatz — APPROVED** [review]
Thanks, @maccabeelevine! I agree that it's a little bit silly-looking, but that's beyond our control when the API does not perform as expected. I agree this is preferable to giving the user some other random type of content, since any attempts at fallback rules would likely turn out to be brittle. Perhaps showing this to EBSCO will help you get an answer about what's going on behind the scenes. In any case, it's definitely better than a blank white screen!

**Pull request merged**